### PR TITLE
F/opt/common subexpressions

### PR DESCRIPTION
**Pull request by 0xNB**
common subexpression optimization

### CODE REVIEW

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+            }
+        }
+
+        @Override
+        public void visit(Conv node) {
+            visitUnary(node, node.getOp(), binding_irnode.ir_opcode.iro_Conv);
+        }
+
+        @Override
+        public void visit(Div node) {
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:266]: Do you consider the memory input of `Div` and `Mod` in any way? The language report does state that behavior of division by zero is undefined, but we can't just ignore the resulting exceptions either (at least so far we haven't).

File: src/test/java/edu/kit/compiler/cmd/EchoLineEndingTest.java [hunk 2]
```diff
@@ -22,15 +23,24 @@
             "\n" +
             "lkjasdfkjalsdf";
 
+    // ignore this test on windows
     @Test
+    @Disabled
```
Comment (g4rgoil) [inline src/test/java/edu/kit/compiler/cmd/EchoLineEndingTest.java:28]: This is not really related to this PR. Also you can just delete the entire file, as it is not doing anything anymore.

File: configure_env [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#!/bin/bash
```
Comment (g4rgoil) [inline configure_env:1]: Do we need this script (same goes for `test`)?

**Comment by 0xNB** [comment]
Replace Unary and Binary Map with Map of Variable Length Keys for function calls etc.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+            // do nothing
+        }
+
+        @Override
+        public void visit(Call call) {
+            visitPreds(call, iterableToArray(call.getPreds()), binding_irnode.ir_opcode.iro_Call);
+        }
+
+        @Override
+        public void visit(Const node) {
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:309]: If you want to eliminate function calls, you also need to handle `Address` nodes in some way. They should be pretty much like `Const`, just using the referenced `Entitiy` as key, instead of the `TargetValue`.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+    private final class CSEVisitor extends NodeVisitor.Default {
+
+        private final Map<TargetValue, Node> constCache = new HashMap<>();
+        private final Map<NodePreds, Node> nodeCache = new HashMap<>();
+
+        @RequiredArgsConstructor
+        private class NodePreds {
+            private final Node[] preds;
+            private final binding_irnode.ir_opcode opcode;
+            private final Mode mode;
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:216]: The `Mode` is only included in your `equals`. This breaks the contract concerning the behavior of `equals` and `hashCode`. Maybe consider just letting `lombok` generate the two methods for you.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+        public void defaultVisit(Node node) {
+            // change nothing
+        }
+
+        private void visitPreds(Node node, Node[] preds, binding_irnode.ir_opcode opcode) {
+            NodePreds np = new NodePreds(preds, opcode, node.getMode());
+            if (nodeCache.containsKey(np)) {
+                Node n = nodeCache.get(np);
+                // can exchange node with already found in list
+                nodeValues.put(node, n);
```
Comment (N-Maas) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:241]: It might be a good idea to check which of `node` / `n` dominates the other (if any) and put the dominating one in the cache.

**Review by N-Maas — COMMENTED** [review]
Generally seems to work at least for nodes in the same block. However, you definitively need to clean up the code. There are print statements and graph dumps at multiple locations.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+        for (BackEdges.Edge edge : BackEdges.getOuts(n)) {
+            Node node = edge.node;
+            if (node.getOpCode() == binding_irnode.ir_opcode.iro_Call) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private int dominates(Pointer source, Pointer target) {
```
Comment (N-Maas) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:174]: This should rather return a boolean.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+        if (newNode != null) {
+            Pointer newPointer = newNode.getBlock().ptr;
+            Pointer origPointer = orig.getBlock().ptr;
+            Pointer replacementPointer = replacement.getBlock().ptr;
+
+            int dominatesOld = dominates(newPointer, origPointer);
+            int dominatesNew = dominates(newPointer, replacementPointer);
+
+            if (dominatesOld == 0 || dominatesNew == 0) {
+                // can not replace node in this case since new node does not dominate one of the replacements
```
[the remark this answers is not in the document]
Comment (0xNB) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:141]: Both replacements are checked since both nodes are inserted to be checked for similarity here.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+            }
+            // new node has to dominate all uses
+            boolean usesDominated = dominatesUses(newNode, orig);
+            boolean usesDominatedRep = dominatesUses(newNode, replacement);
+
+            if (hasCallAsChild(orig) || hasCallAsChild(replacement)) {
+                System.out.println("call as child do not replace");
+                return false;
+            }
+            if (usesDominated && usesDominatedRep) {
```
[the remark this answers is not in the document]
Comment (0xNB) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:152]: `dominates()` checks for block dominance while the use can be in another block that lies before the block of the to be replaced node. This is the case for loops ie.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+
+            if (dominatesOld == 0 || dominatesNew == 0) {
+                // can not replace node in this case since new node does not dominate one of the replacements
+                return false;
+            }
+            // new node has to dominate all uses
+            boolean usesDominated = dominatesUses(newNode, orig);
+            boolean usesDominatedRep = dominatesUses(newNode, replacement);
+
+            if (hasCallAsChild(orig) || hasCallAsChild(replacement)) {
```
[the remark this answers is not in the document]
Comment (0xNB) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:148]: This has been indeed fixed and can be removed here.

File: src/main/java/edu/kit/compiler/optimizations/NodeCollector.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+
+import firm.nodes.Node;
+import firm.nodes.NodeVisitor;
+import lombok.Getter;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Getter
+public class NodeCollector extends NodeVisitor.Default {
```
Comment (N-Maas) [inline src/main/java/edu/kit/compiler/optimizations/NodeCollector.java:11]: This class is currently not used. However, I would suggest to keep it and instead remove `WorklistFiller` and `DefaultWorklist`; the CSE doesn't seem to need a proper worklist anyway (it only iterates over the nodes in the list and doesn't reinsert any nodes).

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+            if (!dominatesOld || !dominatesNew) {
+                // can not replace node if the new node isnt dominating both former nodes.
+                return false;
+            }
+            // new node has to dominate all uses
+            boolean usesDominated = dominatesUses(newNode, orig);
+            boolean usesDominatedRep = dominatesUses(newNode, replacement);
+
+            if (usesDominated && usesDominatedRep) {
+                //      System.out.println("replaced node " + orig + " with " + newNode + " in graph " + g);
```
Comment (N-Maas) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:149]: can be removed

**Review by N-Maas — COMMENTED** [review]
Generally looks good now

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+    /**
+     * Transform the given node with the given associated lattice element if
+     * that element is constant.
+     */
+    private boolean transform(Graph g, Node orig, Node replacement, binding_irnode.ir_opcode opcode) {
+        Node newNode = switch (opcode) {
+            case iro_Add -> {
+                Add add = (Add) orig;
+                yield g.newAdd(orig.getBlock(), add.getLeft(), add.getRight());
+            }
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:87]: You have added a number of new nodes the `CSEVisitor`, don't you need to propagate these changes to `transform`?

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+
+        @Override
+        public void visit(Cond node) {
+            // do nothing
+        }
+
+        @Override
+        public void visit(Proj proj) {
+            // do nothing
+        }
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:278]: I think you need to handle `Proj`s as well. For example, if you eliminate a pair of `Load`s, you'd also want to unify their result projections. Obviously beware that `Proj`s are only equal if they extract the same number.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 15]
```diff
@@ -0,0 +1,387 @@
...
+    }
+
+    private boolean dominates(Pointer source, Pointer target) {
+        int dominates = 0;
+
+        if (!dominateMap.containsKey(source)) {
+            // we have no entry yet, compute both values
+            dominates = binding_irdom.block_dominates(source, target);
+
+            HashMap<Pointer, Integer> cachedResults = new HashMap<>();
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:177]: Are you caching for performance reasons only? If so, note that the implementation of `block_dominates` just dereferences a few pointer, and then compares two integers.

File: src/main/java/edu/kit/compiler/optimizations/NodeCollector.java (hunk 11 above)
Comment (larsk21) [reply src/main/java/edu/kit/compiler/optimizations/NodeCollector.java:11]: There are a number of visitors for filling data structures in `optimizations.Util`. We should probably use those instead.

File: src/main/java/edu/kit/compiler/io/DefaultWorklist.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
+package edu.kit.compiler.io;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class DefaultWorklist<T> implements Worklist<T>{
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/io/DefaultWorklist.java:6]: What is the benefit of this class?

File: src/main/java/edu/kit/compiler/optimizations/WorklistFiller.java [hunk 17]
```diff
@@ -0,0 +1,26 @@
+package edu.kit.compiler.optimizations;
+
+import edu.kit.compiler.io.Worklist;
+import firm.nodes.Node;
+import firm.nodes.NodeVisitor;
+
+/**
+ * Firm node visitor that inserts all visited nodes in the given Worklist.
+ */
+public class WorklistFiller extends NodeVisitor.Default {
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/WorklistFiller.java:10]: Same as for `NodeCollector`, there are visitors for filling data structures in `optimizations.Util`.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 18]
```diff
@@ -0,0 +1,387 @@
...
+                // for each replacement merge nodes together
+                Node orig = replacement.getKey();
+                Node n = replacement.getValue();
+
+                changes |= transform(g, orig, n, orig.getOpCode());
+            }
+            if (changes) {
+                hadChange = true;
+            }
+            maxChanges++;
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:65]: Should this be inside `if (changes)`?

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 19]
```diff
@@ -0,0 +1,387 @@
...
+            }
+            case iro_Minus -> {
+                Minus minus = (Minus) orig;
+                yield g.newMinus(orig.getBlock(), minus.getOp());
+            }
+            case iro_Const -> {
+                Const cons = (Const) orig;
+                yield g.newConst(cons.getTarval().asInt(), cons.getMode());
+            }
+            case iro_Sel -> {
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:116]: I think out optimizations runs after `lower`, so we shouldn't have `Sel` nodes at this point.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java (hunk 15 above)
Comment (larsk21) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:177]: This is correct, Firm caches the dominance information itself.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 20]
```diff
@@ -0,0 +1,387 @@
...
+                    // replace and update cache
+                    nodeValues.put(n, node);
+                    nodeCache.put(np, node);
+                }
+            } else {
+                nodeCache.put(np, node);
+            }
+        }
+
+        private Node[] iterableToArray(Iterable<Node> iterable) {
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:247]: Please add this as `toArray` to `io.CommonUtil` as a generic function similar to `CommonUtil.toList` (I would suggest using a Stream).

File: src/test/java/edu/kit/compiler/optimizations/CommonSubexpressionTest.java [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+    private String surroundWithIO(String program, String expression) {
+        return "System.in.read()\n"
+                + program + "\n"
+                + "System.out.write(" + expression + ")";
+    }
+
+
+    private List<Node> getNodes(Graph graph) {
+        List<Node> nodes = new ArrayList<>();
+        NodeVisitor nodeVisitor = new NodeVisitor.Default() {
```
Comment (larsk21) [inline src/test/java/edu/kit/compiler/optimizations/CommonSubexpressionTest.java:38]: You can use the generic node list filler here.

File: src/test/java/edu/kit/compiler/optimizations/FirmGraphGenerator.java [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+                    } catch (IOException e) {
+                        e.printStackTrace();
+                    }
+                }
+            }
+        }
+    }
+
+    public static void dumpGraph(Graph g) throws IOException {
+        JFirmSingleton.initializeFirmLinux();
```
Comment (larsk21) [inline src/test/java/edu/kit/compiler/optimizations/FirmGraphGenerator.java:64]: You have to assure that this is not called more than once.

File: src/test/java/edu/kit/compiler/optimizations/FirmGraphGenerator.java [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+                Files.move(currentFile.toPath(), targetFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+
+                currentFile.delete();
+                System.out.printf("moved file %s\n", s);
+            }
+        }
+    }
+
+    public static List<Graph> getFirmGraphs(String inputFile) throws IOException {
+        JFirmSingleton.initializeFirmLinux();
```
Comment (larsk21) [inline src/test/java/edu/kit/compiler/optimizations/FirmGraphGenerator.java:88]: see above

File: src/test/java/edu/kit/compiler/optimizations/FirmGraphGenerator.java [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+        try {
+            ast.accept(irv);
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        return StreamSupport.stream(Program.getGraphs().spliterator(), false).collect(Collectors.toList());
+    }
+
+    public static int getGraphCount() {
+        JFirmSingleton.initializeFirmLinux();
```
Comment (larsk21) [inline src/test/java/edu/kit/compiler/optimizations/FirmGraphGenerator.java:125]: see above

File: src/test/java/edu/kit/compiler/optimizations/FirmGraphGenerator.java (hunk 22 above)
Comment (N-Maas) [reply src/test/java/edu/kit/compiler/optimizations/FirmGraphGenerator.java:64]: Actually, this is already the job of the singleton.

Comment (larsk21) [reply src/test/java/edu/kit/compiler/optimizations/FirmGraphGenerator.java:64]: Ah, I see.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (0xNB) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:148]: I think this conversation can also be closed? @N-Maas Or if you have another idea we can use that aswell

File: src/main/java/edu/kit/compiler/optimizations/NodeCollector.java (hunk 11 above)
Comment (0xNB) [reply src/main/java/edu/kit/compiler/optimizations/NodeCollector.java:11]: `DefaultWorklist` and `WorklistFiller` have been replaced with the `NodeCollector` instead

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 25]
```diff
@@ -0,0 +1,387 @@
...
+                Mul mul = (Mul) orig;
+                yield g.newMul(orig.getBlock(), mul.getLeft(), mul.getRight());
+            }
+            case iro_Minus -> {
+                Minus minus = (Minus) orig;
+                yield g.newMinus(orig.getBlock(), minus.getOp());
+            }
+            case iro_Const -> {
+                Const cons = (Const) orig;
+                yield g.newConst(cons.getTarval().asInt(), cons.getMode());
```
[the remark this answers is not in the document]
Comment (0xNB) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:114]: Done

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 26]
```diff
@@ -0,0 +1,387 @@
...
+
+    private boolean hasCallAsChild(Node n) {
+        for (BackEdges.Edge edge : BackEdges.getOuts(n)) {
+            Node node = edge.node;
+            if (node.getOpCode() == binding_irnode.ir_opcode.iro_Call) {
+                return true;
+            }
+        }
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (0xNB) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:168]: Done

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (N-Maas) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:148]: I'm still not sure whether the check is actually necessary, but in doubt the more conservative option is alright

File: src/main/java/edu/kit/compiler/io/DefaultWorklist.java (hunk 16 above)
Comment (0xNB) [reply src/main/java/edu/kit/compiler/io/DefaultWorklist.java:6]: Removed

File: src/main/java/edu/kit/compiler/optimizations/WorklistFiller.java (hunk 17 above)
Comment (0xNB) [reply src/main/java/edu/kit/compiler/optimizations/WorklistFiller.java:10]: Removed

File: src/main/java/edu/kit/compiler/optimizations/NodeCollector.java (hunk 11 above)
Comment (larsk21) [reply src/main/java/edu/kit/compiler/optimizations/NodeCollector.java:11]: Why create the new class `NodeCollector`, if we already have `Util.NodeListFiller`?

File: src/test/java/edu/kit/compiler/optimizations/CommonSubexpressionTest.java [hunk 27]
```diff
@@ -0,0 +1,132 @@
...
+        // return a + d;
+        // ->
+        // int tmp = b * c;
+        // int a = tmp + g;
+        // int d = tmp * e;
+        // return a + d;
+        String file = "edu/kit/compiler/optimizations/Subexpressions.java";
+        Graph g = build(file);
+
+        Dump.dumpGraph(g, "before-opt");
```
Comment (larsk21) [inline src/test/java/edu/kit/compiler/optimizations/CommonSubexpressionTest.java:75]: Please remove all `dumpGraph`s before merging

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 28]
```diff
@@ -0,0 +1,466 @@
...
+        }
+
+        @Override
+        public void visit(Minus node) {
+            visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Minus);
+        }
+
+        @Override
+        public void visit(Mod node) {
+            //   visitPreds(node, Util.iterableToArray(node.getPreds()), binding_irnode.ir_opcode.iro_Mod);
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:413]: `Mod` and `Div` are still commented out.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 29]
```diff
@@ -0,0 +1,466 @@
...
+        g.assureProperties(IR_GRAPH_PROPERTY_CONSISTENT_DOMINANCE);
+
+        return hadChange;
+    }
+
+    /**
+     * Transform the given node with the given associated lattice element if
+     * that element is constant.
+     */
+    private boolean transform(Graph g, Node orig, Node replacement, binding_irnode.ir_opcode opcode, HashMap<Node, Node> replacementMap) {
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:108]: `Div` and `Mod` are also missing from `transform`.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 30]
```diff
@@ -0,0 +1,466 @@
...
+                if (dominates(n.getBlock().ptr, node.getBlock().ptr)) {
+                    nodeValues.put(node, n);
+                } else {
+                    nodeValues.put(n, node);
+                    cmpCache.put(cmpPreds, node);
+                }
+            } else {
+                cmpCache.put(cmpPreds, node);
+            }
+        }
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:338]: You're not actually doing anything with this information. Also, I don't see any value in eliminating `Cmp`s.

Comment (larsk21) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:338]: In addition, `CmpPreds` does not define an `equals` method, so `cmpCache.containsKey(cmdPreds)` will always be `false`.

Comment (0xNB) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:338]: The nodeValues is updated which exchanges `Cmp` nodes, I have added equals and hashCode here though.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java (hunk 28 above)
Comment (0xNB) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:413]: Done

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 31]
```diff
@@ -0,0 +1,466 @@
...
+                Graph.exchange(addr, newAddr);
+            } else {
+                addressCache.put(ent, addr);
+            }
+        }
+
+        @Override
+        public void visit(Call call) {
+            Attributes a = s.getAttributeAnalysis().getAttributes(Util.getCallee(call));
+            if (a.isPure()) {
```
Comment (g4rgoil) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:375]: This check is not sufficient. You either need to check `terminates` (or preferably `Util#isPinned`) and `canDeduplicate`.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 32]
```diff
@@ -0,0 +1,466 @@
...
+            binding_irgopt.remove_bads(g.ptr);
+        } while (changes && maxChanges < 10);
+
+        if (!backEdgesEnabled) {
+            BackEdges.disable(g);
+        }
+        binding_irgopt.remove_bads(g.ptr);
+        binding_irgopt.remove_unreachable_code(g.ptr);
+        binding_irgopt.remove_bads(g.ptr);
+        g.assureProperties(IR_GRAPH_PROPERTY_CONSISTENT_DOMINANCE);
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:99]: Shouldn't this be `confirmProperties(NONE)`?

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 33]
```diff
@@ -0,0 +1,466 @@
...
+            }
+            if (changes) {
+                hadChange = true;
+                maxChanges++;
+            }
+
+            binding_irgopt.remove_bads(g.ptr);
+            binding_irgopt.remove_unreachable_code(g.ptr);
+            binding_irgopt.remove_bads(g.ptr);
+        } while (changes && maxChanges < 10);
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:91]: Is it necessary to have a fix point iteration here where a new `CSEVisitor` is created in every iteration?

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 34]
```diff
@@ -0,0 +1,466 @@
...
+
+            HashMap<Node, Node> replacementMap = new HashMap<>();
+            for (var replacement : nodeValues.entrySet()) {
+                // for each replacement merge nodes together
+                Node orig = replacement.getKey();
+                Node n = replacement.getValue();
+
+                int maxIndirections = 150;
+                while (orig.getOpCode() == binding_irnode.ir_opcode.iro_Deleted && maxIndirections > 0) {
+                    orig = replacementMap.getOrDefault(orig, orig);
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:65]: Given `orig.getOpCode() == iro_Deleted` is `true` and `replacementMap.get(orig)` is empty, then `replacementMap.getOrDefault(orig, orig)` will return `orig` and the condition `orig.getOpCode() == iro_Deleted` will continue to be `true`, right?

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 35]
```diff
@@ -0,0 +1,466 @@
...
+
+            CSEVisitor visitor = new CSEVisitor(state);
+
+            List<Node> nodes = new ArrayList<>();
+            Util.NodeListFiller nodeCollector = new Util.NodeListFiller(nodes);
+            g.walkTopological(nodeCollector);
+            while (!nodes.isEmpty()) {
+                Node n = nodes.remove(0);
+                n.accept(visitor);
+            }
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:55]: Is there a difference to `g.walkTopological(visitor)`?

Comment (g4rgoil) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:54]: There does seem to be a difference. Probably because `Address` and `Const` nodes are replaces in the visitor. That should probably moved to `transform`.

Comment (larsk21) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:54]: Good point, at first look I thought the Visitor was only for analysis

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java (hunk 34 above)
Comment (g4rgoil) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:66]: The default case should never occur (and in practice doesn't seem to), so I would leave this as it is for now. The `maxIndirection` is actually hit a few times.

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 36]
```diff
@@ -0,0 +1,412 @@
...
+            binding_irgopt.remove_bads(g.ptr);
+        } while (changes && numChanges < MAX_CHANGES);
+
+        if (!backEdgesEnabled) {
+            BackEdges.disable(g);
+        }
+
+        binding_irgopt.remove_bads(g.ptr);
+        binding_irgopt.remove_unreachable_code(g.ptr);
+        binding_irgopt.remove_bads(g.ptr);
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:87]: Is this not done before we leave the loop?

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java [hunk 37]
```diff
@@ -0,0 +1,412 @@
...
+            if (!dominates(newNode.getBlock().ptr, n.getBlock().ptr)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static Node findReplacement(Node node, Map<Node, Node> replacements) {
+        for (int i = 0; i < MAX_INDIRECTIONS; ++i) {
+            node = replacements.get(node);
```
Comment (larsk21) [inline src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:229]: `replacements` does always contain the node as key, right?

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java (hunk 36 above)
Comment (g4rgoil) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:87]: Fair point

File: src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java (hunk 37 above)
Comment (g4rgoil) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:229]: In theory, yes `node` _should_ never be `null` ...

Comment (larsk21) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java:229]: And if I loop for a couple of times, replacing `node` with `replacements.get(node)`, I will never find a `node != null` with `replacements.containsKey(node) == false`?

Comment (g4rgoil) [reply src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java]: Actually, this function is not correct as it is right now (see the newest commit). I'm really not sure what the check for `iro_Deleted` is supposed to do.

**Review by larsk21 — APPROVED** [review]
Ideally, we would fix the analysis first, but since this should give us an additional speedup, we should merge it now.

**Comment by larsk21** [comment]
see #174
